photon-ml/src/main/scala/com/linkedin/photon/ml/function/DiffFunction.scala

```diff
@@ -15,307 +15,38 @@
- * the underlying vector hasn't changed.
- *
- * FAQ 2: Why use cumGradient (cumulative gradient)?
- * A: Using cumGradient would allow the functions to modify and then return cumGradient instead of creating a new
- * gradient vector to avoid memory allocation.
- *
- * @tparam Datum Generic type of data point
- * @author xazhang
+ * Trait for a differentiable function.
  */
-
-trait DiffFunction[Datum <: DataPoint] extends Serializable {
-
-  /**
-   * The depth used in treeAggregate. Depth 1 indicates normal linear aggregate.
-   * Using depth > 1 can reduce memory consumption in driver and may also speed up.
-   * It is experimental so far because treeAggregate is unstable in Spark 1.4 and 1.5.
-   */
-  var treeAggregateDepth: Int = 1
-
-  /**
-   * Calculate the gradient of the function given one data point and the model coefficients
-   * @param datum The given datum at which point to compute the gradient
-   * @param coefficients The given model coefficients used to compute the gradient
-   * @return The computed gradient of the function
-   */
-  protected[ml] def gradientAt(datum: Datum, coefficients: Vector[Double]): Vector[Double] = {
-    calculateAt(datum, coefficients)._2
-  }
-
-  /**
-   * Calculate the value of the function given one data point and model coefficients
-   * @param datum The given datum at which point to compute the function's value
-   * @param coefficients The given model coefficients used to compute the function's value
-   * @return The computed value of the function
-   */
-  protected[ml] def valueAt(datum: Datum, coefficients: Vector[Double]): Double = {
-    calculateAt(datum, coefficients)._1
-  }
-
-  /**
-   * Calculate both the value and the gradient of the function given one data point and model coefficients
-   * @param datum The given datum at which point to compute the function's value and gradient
-   * @param coefficients The given model coefficients used to compute the function's value and gradient
-   * @return The computed value and gradient of the function
-   */
-  protected[ml] def calculateAt(datum: Datum, coefficients: Vector[Double]): (Double, Vector[Double]) = {
-    val cumGradient = Utils.initializeZerosVectorOfSameType(coefficients)
-    val value = calculateAt(datum, coefficients, cumGradient)
-    (value, cumGradient)
-  }
-
-  /**
-   * Calculate both the value and the gradient of the function given one data point and model coefficients, with
-   * the computed gradient added to cumGradient in place.
-   * @param datum The given datum at which point to compute the function's value and gradient
-   * @param coefficients The given model coefficients used to compute function's value and gradient
-   * @param cumGradient The cumulative gradient
-   * @return The computed value of the function
-   */
-  protected[ml] def calculateAt(datum: Datum, coefficients: Vector[Double], cumGradient: Vector[Double]): Double
-
+trait DiffFunction extends ObjectiveFunction {
```

Comment:
I wrote "mixin" but I meant "stackable" for the L2 regularization traits. It is the DiffFunction and TwiceDiffFunction which are mixin.